Stripe integration

rlconley opened the pull request:
rebased, working payment

The Stripe secrets are developer sandbox secrets, to be used until general payment testing is complete and it is time to switch to real Stripe settings.

Commit:
Gets basic stripe working

Commit:
Gets basic stripe working

Commit:
Stores Payment. clean ups.

Commit:
Stores Payment. clean ups.

Commit:
Handle DonationLevels on project page & payment errors

Commit:
Handle DonationLevels on project page & payment errors

Commit:
add button for user input amount and metadata hard coded in button for tip

Commit:
add button for user input amount and metadata hard coded in button for tip

Commit:
add tip input to front end

Commit:
add tip input to front end

Commit:
refactor metadata

Commit:
refactor metadata

Commit:
add creation of tip in stripe_callback

Commit:
add creation of tip in stripe_callback

nmashton committed:
Merge branch 'develop' of https://github.com/caktus/revolv into stripe-integration

Commit:
Merge branch 'stripe-integration' of https://github.com/caktus/revolv into stripe-integration

nmashton committed:
Adding event listeners to project pay module

nmashton committed:
Adding Stripe styling and input dependencies

Commit:
add tip to back end

Commit:
removed merge conflicts

Commit:
update project.html to Neil's version

Commit:
working donate button with tip

Commit:
removed hard coded stripe keys

nmashton committed:
Beautifying

nmashton committed:
Merge

trawick commented on revolv/payments/migrations/0023_tip.py:
```diff
@@ -0,0 +1,28 @@
+                ('id', models.AutoField(verbose_name='ID', serialize=False, auto_created=True, primary_key=True)),
+                ('timestamp', models.DateTimeField(auto_now_add=True)),
+                ('amount', models.FloatField()),
+                ('user', models.ForeignKey(to=settings.AUTH_USER_MODEL)),
```
Regenerate migration based on new FK type

rlconley replied:
This is done.  Migration 0025

trawick commented on revolv/payments/models.py:
```diff
@@ -528,3 +539,12 @@ class ProjectMontlyRepaymentConfig(models.Model):
+    """
+    timestamp = models.DateTimeField(auto_now_add=True, blank=True)
+    user = models.ForeignKey('base.RevolvUserProfile')
+    amount = models.FloatField()
```
Add simple string formatter, like

```
    def __unicode__(self):
        return 'Tip of %s from %s at %s' % (self.amount, self.user, self.timestamp)
```

Test it, and enable a useful feature anyway, by adding this to admin.py in this app:

```
(add Tip to import at top)

admin.site.register(Tip)
```

Now these can be watched in admin.

rlconley replied:
Added and looked at it in admin.

trawick commented on revolv/settings/base-orig.py:
```diff
@@ -176,9 +176,9 @@
-        'HOST': '127.0.0.1',
+        'USER': '',
+        'PASSWORD': '',
+        'HOST': '',
```
If this file isn't ever used, I'd avoid modifying it.

rlconley replied:
That's interesting.  I didn't intentionally modify it.  Will change it back.

rlconley replied:
Changed.

trawick commented on revolv/settings/base-orig.py:
```diff
@@ -462,3 +463,7 @@
+
+#Stripe
+STRIPE_SECRET_KEY = 'sk_test_NoHD98Ut8uIJYbzfgisPVI9N'
+STRIPE_PUBLISHABLE = 'pk_test_W7T0xyvfKErlPJlBjZBCs6VW'
```
pull this out, pick it up from environment with default of ''

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
 
+def stripe_callback(request, pk):
+    import stripe
+    stripe.api_key = settings.STRIPE_SECRET_KEY
```
Unless there is something weird about this module, I think you should import stripe in the third-party imports section above, and set stripe.api_key at global scope, just above this function

rlconley replied:
Makes sense.  Will do.

rlconley replied:
done.

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+from revolv.project.models import Category, Project, ProjectUpdate, DonationLevel
 from revolv.tasks.sfdc import send_donation_info
 
+def stripe_callback(request, pk):
```
not code you wrote, but I'd like to see the name of this changed...  "stripe_callback" will make many people think "this is the code that Stripe calls after a payment" but actually this is just the view which handles the form post; it could be called simply `stripe_payment`.

trawick replied:
silly I know, but this function needs another blank line in front :)

rlconley replied:
Space fixed, method changed to stripe_payment

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+        project=project,
+        payment_type=PaymentType.objects.get_stripe(),
+    )
+    print payment
```
time to zap all these print statements; feel free to write a debug log message with any critical info for debugging in the future

rlconley replied:
removed the print statements that I put in

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+        #log it
+        msg = "Payment error. Re-volv has been notified."
+        return render(request, "project/project_donate_error.html", {"msg": msg, "project": project})
+        pass
```
`pass` is never reached

rlconley replied:
Should be out-dented?

trawick replied:
I think `pass` just gets deleted.

Here's one way to collect the error messages from different exceptions and return to the user regardless of the exception type:

```
diff --git a/revolv/project/views.py b/revolv/project/views.py
index 93df1a5..5af2169 100644
--- a/revolv/project/views.py
+++ b/revolv/project/views.py
@@ -1,4 +1,5 @@
 from decimal import Decimal
+import logging

 from django.conf import settings
 from django.contrib import messages
@@ -21,6 +22,10 @@ from revolv.project import forms
 from revolv.project.models import Category, Project, ProjectUpdate, DonationLevel
 from revolv.tasks.sfdc import send_donation_info

+
+logger = logging.getLogger(__name__)
+
+
 def stripe_callback(request, pk):
     import stripe
     stripe.api_key = settings.STRIPE_SECRET_KEY
@@ -32,18 +37,21 @@ def stripe_callback(request, pk):
     amount_cents = int(request.POST['amount_cents'])
     donation_cents = amount_cents - tip_cents

+    error_msg = None
     try:
-        charge = stripe.Charge.create(source=request.POST['stripeToken'], currency="usd", amount=amount_cents)
-        print request.POST
+        stripe.Charge.create(source=request.POST['stripeToken'], currency="usd", amount=amount_cents)
     except stripe.error.CardError, e:
-        msg = body['error']['message']
+        body = e.json_body
+        error_msg = body['error']['message']
     except stripe.error.APIConnectionError, e:
-        msg = body['error']['message']
+        body = e.json_body
+        error_msg = body['error']['message']
     except Exception, e:
-        #log it
-        msg = "Payment error. Re-volv has been notified."
-        return render(request, "project/project_donate_error.html", {"msg": msg, "project": project})
-        pass
+        error_msg = "Payment error. Re-volv has been notified."
+        logger.exception(error_msg)
+
+    if error_msg:
+        return render(request, "project/project_donate_error.html", {"msg": error_msg, "project": project})

     payment = Payment.objects.create(
         user=request.user.revolvuserprofile,
```

There are some hints at https://stripe.com/docs/api/python#errors (search for "Handling errors").  The several different exception types in this code need to be checked.  I guess one happens with a bad card number, another happens with turning off your network so that creating a Charge can't contact stripe; maybe corrupting the token would fall into the general Exception case???

rlconley replied:
Refactored this.

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+    except Exception, e:
+        #log it
+        msg = "Payment error. Re-volv has been notified."
+        return render(request, "project/project_donate_error.html", {"msg": msg, "project": project})
```
this `render` isn't reached for the stripe.error exceptions handled previously; those need to fall into the error response

rlconley replied:
Addressed in refactor.

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+def stripe_callback(request, pk):
+    import stripe
+    stripe.api_key = settings.STRIPE_SECRET_KEY
+    token = stripe.Token.retrieve(request.POST['stripeToken'])
```
if I understand correctly, we never use token after this call; is there error checking missing here?

rlconley replied:
Yes.  It is missing.

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+    stripe.api_key = settings.STRIPE_SECRET_KEY
+    token = stripe.Token.retrieve(request.POST['stripeToken'])
+
+    project = get_object_or_404(Project, pk=pk)
```
move this to before the first access of POST arguments, to effectively separate handling of parameters vs. handling of POST data

rlconley replied:
Moved.

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+    tip_cents = int(request.POST['metadata'])
+    amount_cents = int(request.POST['amount_cents'])
+    donation_cents = amount_cents - tip_cents
+
```
validation of POST arguments (presence and values) would be useful here  make sure donation_cents <= 0 is caught

I guess the quick and dirty way looks something like this:

```
    try:
        stripe_token = request.POST['stripeToken']
        tip_cents = request.POST['metadata']
        amount_cents = request.POST['amount_cents']
    except KeyError:
        log a message, return invalid request error

    try:
        tip_cents = int(tip_cents)
        amount_cents = int(amount_cents])
        if not (0 < amount_cents < MAX_PAYMENT_CENTS):
            raise ValueError('amount_cents cannot be negative')
        if not (0 <= tip_cents < MAX_PAYMENT_CENTS):
            raise ValueError('tip_cents cannot be negative')
    except ValueError:
        log message, return invalid request error
```

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+        charge = stripe.Charge.create(source=request.POST['stripeToken'], currency="usd", amount=amount_cents)
+        print request.POST
+    except stripe.error.CardError, e:
+        msg = body['error']['message']
```
I guess this is from some sample code; we don't have `body` here AFAICT, so maybe you+Alex can find the sample this came from and figure out what `body` should be

trawick commented on revolv/project/views.py:
```diff
@@ -14,12 +15,53 @@
+    donation_cents = amount_cents - tip_cents
+
+    try:
+        charge = stripe.Charge.create(source=request.POST['stripeToken'], currency="usd", amount=amount_cents)
```
`charge` never used; does the `.create` call do everything we need?

rlconley replied:
Took out "charge"

Commit:
fixes based on codereview of pull request

Commit:
refactored error handling and changed name of of payment method in url

trawick committed:
Error handling improvements

trawick committed:
Use a Stripe sandbox account on production initially

trawick committed:
Force login on stripe_payment() view, log exception for odd Stripe errors

nmashton committed:
Enforcing input restrictions

nmashton committed:
Disabling "+"

trawick committed:
Update PaymentType test for inclusion of _STRIPE

trawick committed:
Basic tests for Stripe payments

The lack of trailing slash hurt an assertRedirects since
even the initial redirect (without trailing slash) was
redirected.

trawick committed:
Merge branch 'stripe-integration' of github.com:caktus/revolv into stripe-integration

trawick committed:
Merge branch 'develop' of github.com:caktus/revolv into stripe-integration

trawick committed:
back out changes to base-orig.py

trawick committed:
space after hash-sign

Commit:
added donation levels to template

trawick committed:
allow rollback

trawick committed:
Merge branch 'stripe-integration' of github.com:caktus/revolv into stripe-integration

nmashton committed:
Code for pre-set donation level front-end

trawick commented:
+1
